lib/dpctl.c

```diff
@@ -1041,24 +1043,41 @@ dpctl_dump_flows(int argc, const char *argv[], struct dpctl_params *dpctl_p)
             }
             minimatch_destroy(&minimatch);
         }
-        ds_clear(&ds);
-        /* If 'pmd_id' is specified, overlapping flows could be dumped from
-         * different pmd threads.  So, separates dumps from different pmds
-         * by printing a title line. */
-        if (pmd_id != f.pmd_id) {
-            if (f.pmd_id == NON_PMD_CORE_ID) {
-                ds_put_format(&ds, "flow-dump from non-dpdk interfaces:\n");
+
+        if (dpctl_p->counters) {
+            if (f.attrs.offloaded) {
+                nb_offloaded++;
             } else {
-                ds_put_format(&ds, "flow-dump from pmd on cpu core: %d\n",
-                              f.pmd_id);
+                nb_ovs++;
+            }
+        } else {
+            ds_clear(&ds);
+            /* If 'pmd_id' is specified, overlapping flows could be dumped from
+             * different pmd threads.  So, separates dumps from different pmds
+             * by printing a title line. */
+            if (pmd_id != f.pmd_id) {
+                if (f.pmd_id == NON_PMD_CORE_ID) {
+                    ds_put_format(&ds, "flow-dump from non-dpdk interfaces:\n");
+                } else {
+                    ds_put_format(&ds, "flow-dump from pmd on cpu core: %d\n",
+                                  f.pmd_id);
+                }
+                pmd_id = f.pmd_id;
+            }
+            if (flow_passes_type_filter(&f, &dump_types)) {
+                format_dpif_flow(&ds, &f, portno_names, dpctl_p);
+                dpctl_print(dpctl_p, "%s\n", ds_cstr(&ds));
             }
-            pmd_id = f.pmd_id;
-        }
-        if (flow_passes_type_filter(&f, &dump_types)) {
-            format_dpif_flow(&ds, &f, portno_names, dpctl_p);
-            dpctl_print(dpctl_p, "%s\n", ds_cstr(&ds));
         }
     }
+
+    if (dpctl_p->counters) {
+        ds_clear(&ds);
+        ds_put_format(&ds, "flows statistics: ovs %u, offloaded %u\n",
```

Comment:
This might confuse users, because ovs-dpctl utility is not able to see offloaded flows.
There are few options if the call was not from appctl:
- print a warning here like "ovs-dpctl is not able to retrieve number of offloaded flows. Use 'ovs-appctl dpctl/dump-flows' instead."
- don't print 'offloaded' statistics in this case.
- print "unavailable" or "unknown" instead of nb_offloaded.

Reply:
I think the ovs-dpctl utility be able to see offloaded flows, the display of commands as below. I can only display the none-zero numbers of flows statistics.

If the offload flows statistics is zero, the display as below:
```bash
ovs-dpctl dump-flows --counters
flows statistics: ovs 4.
```
or
```bash
ovs-dpctl dump-flows --counters
flows statistics: offloaded 4.
```

```bash
# ovs-dpctl dump-flows type=offloaded
in_port(2),eth(src=50:6b:4b:87:ed:1a,dst=01:80:c2:00:00:0e),eth_type(0x88cc), packets:0, bytes:0, used:8.370s, actions:drop
in_port(2),eth(src=50:6b:4b:87:ed:1a,dst=f2:12:37:02:74:a6),eth_type(0x0800),ipv4(frag=no), packets:142777, bytes:9442818, used:0.490s, actions:3
in_port(3),eth(src=f2:12:37:02:74:a6,dst=50:6b:4b:87:ed:1a),eth_type(0x0800),ipv4(frag=no), packets:6327505, bytes:9577799602, used:0.490s, actions:2

# ovs-dpctl dump-flows
in_port(3),eth(src=f2:12:37:02:74:a6,dst=50:6b:4b:87:ed:1a),eth_type(0x0800),ipv4(frag=no), packets:12319629, bytes:18648348418, used:0.530s, actions:2
in_port(2),eth(src=50:6b:4b:87:ed:1a,dst=f2:12:37:02:74:a6),eth_type(0x0800),ipv4(frag=no), packets:278798, bytes:18420204, used:0.530s, actions:3
recirc_id(0),in_port(2),eth(src=50:6b:4b:87:ed:1a,dst=ff:ff:ff:ff:ff:ff),eth_type(0x0806),arp(sip=4.4.4.1,tip=4.4.4.2,op=1/0xff), packets:0, bytes:0, used:never, actions:51,32,14,45,42,39,22,21,27,46,11,64,44,47,13,48,8,36,3,5,40,37,26,61,17,15,34,54,58,10,7,55,19,4,18,12,53,59,43,63,57,25,65,6,52,38,41,20,60,49,1,24,35,16,56,9,23,50,33,30,29,62,31,28
recirc_id(0),in_port(2),eth(src=50:6b:4b:87:ed:1a,dst=ff:ff:ff:ff:ff:ff),eth_type(0x0806),arp(sip=4.4.4.1,tip=4.4.4.3,op=1/0xff), packets:9, bytes:540, used:7.564s, actions:51,32,14,45,42,39,22,21,27,46,11,64,44,47,13,48,8,36,3,5,40,37,26,61,17,15,34,54,58,10,7,55,19,4,18,12,53,59,43,63,57,25,65,6,52,38,41,20,60,49,1,24,35,16,56,9,23,50,33,30,29,62,31,28
recirc_id(0),in_port(2),eth(src=50:6b:4b:87:ed:1a,dst=33:33:00:00:00:01),eth_type(0x86dd),ipv6(frag=no), packets:27, bytes:2106, used:0.066s, actions:51,32,14,45,42,39,22,21,27,46,11,64,44,47,13,48,8,36,3,5,40,37,26,61,17,15,34,54,58,10,7,55,19,4,18,12,53,59,43,63,57,25,65,6,52,38,41,20,60,49,1,24,35,16,56,9,23,50,33,30,29,62,31,28

# ovs-dpctl dump-flows -m
ufid:d7103abd-36b1-4d23-81a8-1c0f9de0452a, skb_priority(0/0),skb_mark(0/0),in_port(enp66s0f0_0),packet_type(ns=0/0,id=0/0),eth(src=f2:12:37:02:74:a6,dst=50:6b:4b:87:ed:1a),eth_type(0x0800),ipv4(src=0.0.0.0/0.0.0.0,dst=0.0.0.0/0.0.0.0,proto=0/0,tos=0/0,ttl=0/0,frag=no), packets:17199891, bytes:26035231047, used:0.150s, offloaded:yes, dp:tc, actions:enp66s0f0
ufid:a72a1706-4126-4b80-b54e-8c905daafeba, skb_priority(0/0),skb_mark(0/0),in_port(enp66s0f0),packet_type(ns=0/0,id=0/0),eth(src=50:6b:4b:87:ed:1a,dst=f2:12:37:02:74:a6),eth_type(0x0800),ipv4(src=0.0.0.0/0.0.0.0,dst=0.0.0.0/0.0.0.0,proto=0/0,tos=0/0,ttl=0/0,frag=no), packets:390123, bytes:25767599, used:0.150s, offloaded:yes, dp:tc, actions:enp66s0f0_0
ufid:7bea0851-591a-427c-ad00-ee0e68554016, recirc_id(0),dp_hash(0/0),skb_priority(0/0),in_port(enp66s0f0),skb_mark(0/0),ct_state(0/0),ct_zone(0/0),ct_mark(0/0),ct_label(0/0),eth(src=50:6b:4b:87:ed:1a,dst=ff:ff:ff:ff:ff:ff),eth_type(0x0806),arp(sip=4.4.4.1,tip=4.4.4.2,op=1/0xff,sha=00:00:00:00:00:00/00:00:00:00:00:00,tha=00:00:00:00:00:00/00:00:00:00:00:00), packets:4, bytes:240, used:0.348s, dp:ovs, actions:enp66s0f0_48,enp66s0f0_29,enp66s0f0_11,enp66s0f0_40,enp66s0f0_39,enp66s0f0_36,enp66s0f0_19,enp66s0f0_18,enp66s0f0_26,enp66s0f0_44,enp66s0f0_8,enp66s0f0_61,enp66s0f0_41,enp66s0f0_42,enp66s0f0_9,enp66s0f0_45,enp66s0f0_7,enp66s0f0_34,enp66s0f0_0,enp66s0f0_2,enp66s0f0_38,enp66s0f0_33,enp66s0f0_24,enp66s0f0_58,enp66s0f0_14,enp66s0f0_12,enp66s0f0_30,enp66s0f0_51,enp66s0f0_62,enp66s0f0_6,enp66s0f0_4,enp66s0f0_53,enp66s0f0_16,enp66s0f0_1,enp66s0f0_13,enp66s0f0_10,enp66s0f0_50,enp66s0f0_60,enp66s0f0_43,enp66s0f0_59,enp66s0f0_54,enp66s0f0_22,enp66s0f0_55,enp66s0f0_3,enp66s0f0_49,enp66s0f0_31,enp66s0f0_37,enp66s0f0_17,enp66s0f0_56,enp66s0f0_46,cni0,enp66s0f0_21,enp66s0f0_32,enp66s0f0_15,enp66s0f0_52,enp66s0f0_5,enp66s0f0_20,enp66s0f0_47,enp66s0f0_35,enp66s0f0_27,enp66s0f0_25,enp66s0f0_57,enp66s0f0_28,enp66s0f0_23
ufid:8a0e90b5-55db-41f7-95be-4e1692ba451b, recirc_id(0),dp_hash(0/0),skb_priority(0/0),in_port(enp66s0f0),skb_mark(0/0),ct_state(0/0),ct_zone(0/0),ct_mark(0/0),ct_label(0/0),eth(src=50:6b:4b:87:ed:1a,dst=ff:ff:ff:ff:ff:ff),eth_type(0x0806),arp(sip=4.4.4.1,tip=4.4.4.3,op=1/0xff,sha=00:00:00:00:00:00/00:00:00:00:00:00,tha=00:00:00:00:00:00/00:00:00:00:00:00), packets:0, bytes:0, used:never, dp:ovs, actions:enp66s0f0_48,enp66s0f0_29,enp66s0f0_11,enp66s0f0_40,enp66s0f0_39,enp66s0f0_36,enp66s0f0_19,enp66s0f0_18,enp66s0f0_26,enp66s0f0_44,enp66s0f0_8,enp66s0f0_61,enp66s0f0_41,enp66s0f0_42,enp66s0f0_9,enp66s0f0_45,enp66s0f0_7,enp66s0f0_34,enp66s0f0_0,enp66s0f0_2,enp66s0f0_38,enp66s0f0_33,enp66s0f0_24,enp66s0f0_58,enp66s0f0_14,enp66s0f0_12,enp66s0f0_30,enp66s0f0_51,enp66s0f0_62,enp66s0f0_6,enp66s0f0_4,enp66s0f0_53,enp66s0f0_16,enp66s0f0_1,enp66s0f0_13,enp66s0f0_10,enp66s0f0_50,enp66s0f0_60,enp66s0f0_43,enp66s0f0_59,enp66s0f0_54,enp66s0f0_22,enp66s0f0_55,enp66s0f0_3,enp66s0f0_49,enp66s0f0_31,enp66s0f0_37,enp66s0f0_17,enp66s0f0_56,enp66s0f0_46,cni0,enp66s0f0_21,enp66s0f0_32,enp66s0f0_15,enp66s0f0_52,enp66s0f0_5,enp66s0f0_20,enp66s0f0_47,enp66s0f0_35,enp66s0f0_27,enp66s0f0_25,enp66s0f0_57,enp66s0f0_28,enp66s0f0_23
ufid:4bd6e591-8d2b-4f57-b9ea-d6962949ffcb, recirc_id(0),dp_hash(0/0),skb_priority(0/0),in_port(enp66s0f0),skb_mark(0/0),ct_state(0/0),ct_zone(0/0),ct_mark(0/0),ct_label(0/0),eth(src=50:6b:4b:87:ed:1a,dst=33:33:00:00:00:01),eth_type(0x86dd),ipv6(src=::/::,dst=::/::,label=0/0,proto=0/0,tclass=0/0,hlimit=0/0,frag=no), packets:27, bytes:2106, used:3.688s, dp:ovs, actions:enp66s0f0_48,enp66s0f0_29,enp66s0f0_11,enp66s0f0_40,enp66s0f0_39,enp66s0f0_36,enp66s0f0_19,enp66s0f0_18,enp66s0f0_26,enp66s0f0_44,enp66s0f0_8,enp66s0f0_61,enp66s0f0_41,enp66s0f0_42,enp66s0f0_9,enp66s0f0_45,enp66s0f0_7,enp66s0f0_34,enp66s0f0_0,enp66s0f0_2,enp66s0f0_38,enp66s0f0_33,enp66s0f0_24,enp66s0f0_58,enp66s0f0_14,enp66s0f0_12,enp66s0f0_30,enp66s0f0_51,enp66s0f0_62,enp66s0f0_6,enp66s0f0_4,enp66s0f0_53,enp66s0f0_16,enp66s0f0_1,enp66s0f0_13,enp66s0f0_10,enp66s0f0_50,enp66s0f0_60,enp66s0f0_43,enp66s0f0_59,enp66s0f0_54,enp66s0f0_22,enp66s0f0_55,enp66s0f0_3,enp66s0f0_49,enp66s0f0_31,enp66s0f0_37,enp66s0f0_17,enp66s0f0_56,enp66s0f0_46,cni0,enp66s0f0_21,enp66s0f0_32,enp66s0f0_15,enp66s0f0_52,enp66s0f0_5,enp66s0f0_20,enp66s0f0_47,enp66s0f0_35,enp66s0f0_27,enp66s0f0_25,enp66s0f0_57,enp66s0f0_28,enp66s0f0_23
```

Reply:
I modify the code, it'll don't print 'offloaded' statistics. Thanks